lib/licensed/sources/nuget.rb

```diff
@@ -0,0 +1,230 @@
+# frozen_string_literal: true
+require "json"
+require "reverse_markdown"
+
+module Licensed
+  module Sources
+    # Only supports ProjectReference (project.assets.json) style restore used in .NET Core.
+    # Does not currently support packages.config style restore.
+    class NuGet < Source
+      def self.type
+        "nuget"
+      end
+
+      class NuGetDependency < Licensed::Dependency
+        LICENSE_FILE_REGEX = /<license\s*type\s*=\s*\"\s*file\s*\"\s*>\s*(.*)\s*<\/license>/ix.freeze
+        LICENSE_URL_REGEX = /<licenseUrl>\s*(.*)\s*<\/licenseUrl>/ix.freeze
+        PROJECT_URL_REGEX = /<projectUrl>\s*(.*)\s*<\/projectUrl>/ix.freeze
+
+        def initialize(name:, version:, path:, search_root: nil, metadata: {}, errors: [])
+          super(name: name, version: version, path: path, search_root: search_root, metadata: metadata, errors: errors)
+          @metadata["homepage"] = project_url if project_url
+        end
+
+        def nuspec_path
+          name = @metadata["name"]
+          File.join(self.path, "#{name}.nuspec")
+        end
+
+        def nuspec_contents
+          return unless nuspec_path
+          @nuspec_contents ||= File.read(nuspec_path)
+        end
+
+        def project_url
+          return unless nuspec_contents
+          @project_url ||= begin
+            match = nuspec_contents.match PROJECT_URL_REGEX
+            match[1] if match && match[1]
+          end
```

Comment:
The caching here will fail if the evaluated block returns nil.  If that happens, follow up calls to this expand to `@project_url = nil || begin`, which will re-execute the block.

This can be fixed by checking whether `@project_url` has been defined instead of relying on `||=` 👍 

```suggestion
          return @project_url if defined?(@project_url)
          @project_url = begin
            match = nuspec_contents.match PROJECT_URL_REGEX
            match[1] if match && match[1]
          end
```

Reply:
Makes sense, thanks!